IconTreeShakerException: Invalid ConstFinder result

Working on a new widget in the widgets layer, I defined a const Icon using the ascii value for the IconData:

```dart
const Icon(IconData(0x25BA));
```

Since the Icon widget is defined in the widgets layer, I thought this should work fine. Running an app with this builds and works fine.

When I pushed a PR, CI was very unhappy. This use of Icon results in an error from the IconTreeShaker.

```
flutter build apk --release

Target aot_android_asset_bundle failed: IconTreeShakerException: Invalid ConstFinder result. Expected "fontPackage" to be a String, "fontFamily" to be a String, and "codePoint" to be an int, got: {codePoint: 9658, fontFamily: null, fontPackage: null, matchTextDirection: false, fontFamilyFallback: null}.

To disable icon tree shaking, pass --no-tree-shake-icons to the requested flutter build command


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/katelovett/_github/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 14s
Running Gradle task 'assembleRelease'...                           14.7s
Gradle task assembleRelease failed with exit code 1
```

I figured I was holding it wrong and was not allowed to do this, but @goderbauer suspects this is a bug in the icon tree shaker script.

This can be reproduced with the following code and the command `flutter build apk --release`

```dart
import 'package:flutter/material.dart';

void main() => runApp(const SliverTreeExampleApp());

class SliverTreeExampleApp extends StatelessWidget {
  const SliverTreeExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: SliverTreeExample(),
    );
  }
}

class SliverTreeExample extends StatelessWidget {
  const SliverTreeExample({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(child: Icon(IconData(0x25BA)),),
    );
  }
}

```

This is an example build where I saw this in CI: https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20build_tests_2_3/33418/overview

--- a/packages/flutter_tools/lib/src/build_system/targets/icon_tree_shaker.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/icon_tree_shaker.dart
@@ -336,7 +336,7 @@ class IconTreeShaker {
           'got: $iconDataMap.');
       }
       if (fontFamily == null) {
-        _logger.printStatus(
+        _logger.printTrace(
           'Expected to find fontFamily for constant IconData with codepoint: '
           '$codePoint, but found fontFamily: $fontFamily. This usually means '
           'you are relying on the system font. Alternatively, font families in '

--- a/packages/flutter_tools/lib/src/build_system/targets/icon_tree_shaker.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/icon_tree_shaker.dart
@@ -327,15 +327,25 @@ class IconTreeShaker {
       final Object? package = iconDataMap['fontPackage'];
       final Object? fontFamily = iconDataMap['fontFamily'];
       final Object? codePoint = iconDataMap['codePoint'];
-      if ((package ?? '') is! String || // Null is ok here.
-          fontFamily is! String ||
+      if ((package ?? '') is! String ||
+          (fontFamily ?? '') is! String ||
           codePoint is! num) {
         throw IconTreeShakerException._(
           'Invalid ConstFinder result. Expected "fontPackage" to be a String, '
           '"fontFamily" to be a String, and "codePoint" to be an int, '
           'got: $iconDataMap.');
       }
-      final String family = fontFamily;
+      if (fontFamily == null) {
+        _logger.printStatus(
+          'Expected to find fontFamily for constant IconData with codepoint: '
+          '$codePoint, but found fontFamily: $fontFamily. This usually means '
+          'you are relying on the system font. Alternatively, font families in '
+          'an IconData class can be provided in the assets section of your '
+          'pubspec.yaml, or you are missing "uses-material-design: true".',
+        );
+        continue;
+      }
+      final String family = fontFamily as String;
       final String key = package == null
         ? family
         : 'packages/$package/$family';

--- a/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
@@ -561,6 +561,86 @@ void main() {
     expect(processManager, hasNoRemainingExpectations);
   });
 
+  testWithoutContext('Allows system font fallback when there are none', () async {
+    final Environment environment = createEnvironment(<String, String>{
+      kIconTreeShakerFlag: 'true',
+      kBuildMode: 'release',
+    });
+    final File appDill = environment.buildDir.childFile('app.dill')
+      ..createSync(recursive: true);
+
+    // Valid manifest, just not using it.
+    fontManifestContent = DevFSStringContent(validFontManifestJson);
+
+    final IconTreeShaker iconTreeShaker = IconTreeShaker(
+      environment,
+      fontManifestContent,
+      logger: logger,
+      processManager: processManager,
+      fileSystem: fileSystem,
+      artifacts: artifacts,
+      targetPlatform: TargetPlatform.android,
+    );
+
+    addConstFinderInvocation(appDill.path, stdout: emptyConstFinderResult);
+    // Does not throw
+    await iconTreeShaker.subsetFont(
+      input: fileSystem.file(inputPath),
+      outputPath: outputPath,
+      relativePath: relativePath,
+    );
+
+    expect(
+      logger.statusText,
+      'Expected to find fontFamily for constant IconData with codepoint: '
+      '59470, but found fontFamily: null. This usually means '
+      'you are relying on the system font. Alternatively, font families in '
+      'an IconData class can be provided in the assets section of your '
+      'pubspec.yaml, or you are missing "uses-material-design: true".\n',
+    );
+    expect(processManager, hasNoRemainingExpectations);
+  });
+
+  testWithoutContext('Allows system font fallback when there are none', () async {
+    final Environment environment = createEnvironment(<String, String>{
+      kIconTreeShakerFlag: 'true',
+      kBuildMode: 'release',
+    });
+    final File appDill = environment.buildDir.childFile('app.dill')
+      ..createSync(recursive: true);
+
+    // Nothing in font manifest
+    fontManifestContent = DevFSStringContent(emptyFontManifestJson);
+
+    final IconTreeShaker iconTreeShaker = IconTreeShaker(
+      environment,
+      fontManifestContent,
+      logger: logger,
+      processManager: processManager,
+      fileSystem: fileSystem,
+      artifacts: artifacts,
+      targetPlatform: TargetPlatform.android,
+    );
+
+    addConstFinderInvocation(appDill.path, stdout: emptyConstFinderResult);
+    // Does not throw
+    await iconTreeShaker.subsetFont(
+      input: fileSystem.file(inputPath),
+      outputPath: outputPath,
+      relativePath: relativePath,
+    );
+
+    expect(
+      logger.statusText,
+      'Expected to find fontFamily for constant IconData with codepoint: '
+      '59470, but found fontFamily: null. This usually means '
+      'you are relying on the system font. Alternatively, font families in '
+      'an IconData class can be provided in the assets section of your '
+      'pubspec.yaml, or you are missing "uses-material-design: true".\n',
+    );
+    expect(processManager, hasNoRemainingExpectations);
+  });
+
   testWithoutContext('ConstFinder non-zero exit', () async {
     final Environment environment = createEnvironment(<String, String>{
       kIconTreeShakerFlag: 'true',
@@ -601,6 +681,20 @@ const String validConstFinderResult = '''
     {
       "codePoint": 59470,
       "fontFamily": "MaterialIcons",
+      "fontPackage": null,
+      "matchTextDirection": false
+    }
+  ],
+  "nonConstantLocations": []
+}
+''';
+
+const String emptyConstFinderResult = '''
+{
+  "constantInstances": [
+    {
+      "codePoint": 59470,
+      "fontFamily": null,
       "fontPackage": null,
       "matchTextDirection": false
     }
@@ -668,3 +762,5 @@ const String invalidFontManifestJson = '''
   ]
 }
 ''';
+
+const String emptyFontManifestJson = '[]';

--- a/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
@@ -561,7 +561,7 @@ void main() {
     expect(processManager, hasNoRemainingExpectations);
   });
 
-  testWithoutContext('Allows system font fallback when there are none', () async {
+  testWithoutContext('Allow system font fallback when fontFamily is null', () async {
     final Environment environment = createEnvironment(<String, String>{
       kIconTreeShakerFlag: 'true',
       kBuildMode: 'release',
@@ -601,7 +601,7 @@ void main() {
     expect(processManager, hasNoRemainingExpectations);
   });
 
-  testWithoutContext('Allows system font fallback when there are none', () async {
+  testWithoutContext('Allow system font fallback when fontFamily is null and manifest is empty', () async {
     final Environment environment = createEnvironment(<String, String>{
       kIconTreeShakerFlag: 'true',
       kBuildMode: 'release',

--- a/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/icon_tree_shaker_test.dart
@@ -591,12 +591,14 @@ void main() {
     );
 
     expect(
-      logger.statusText,
-      'Expected to find fontFamily for constant IconData with codepoint: '
-      '59470, but found fontFamily: null. This usually means '
-      'you are relying on the system font. Alternatively, font families in '
-      'an IconData class can be provided in the assets section of your '
-      'pubspec.yaml, or you are missing "uses-material-design: true".\n',
+      logger.traceText,
+      contains(
+        'Expected to find fontFamily for constant IconData with codepoint: '
+        '59470, but found fontFamily: null. This usually means '
+        'you are relying on the system font. Alternatively, font families in '
+        'an IconData class can be provided in the assets section of your '
+        'pubspec.yaml, or you are missing "uses-material-design: true".\n'
+      ),
     );
     expect(processManager, hasNoRemainingExpectations);
   });
@@ -631,12 +633,14 @@ void main() {
     );
 
     expect(
-      logger.statusText,
-      'Expected to find fontFamily for constant IconData with codepoint: '
-      '59470, but found fontFamily: null. This usually means '
-      'you are relying on the system font. Alternatively, font families in '
-      'an IconData class can be provided in the assets section of your '
-      'pubspec.yaml, or you are missing "uses-material-design: true".\n',
+      logger.traceText,
+      contains(
+        'Expected to find fontFamily for constant IconData with codepoint: '
+        '59470, but found fontFamily: null. This usually means '
+        'you are relying on the system font. Alternatively, font families in '
+        'an IconData class can be provided in the assets section of your '
+        'pubspec.yaml, or you are missing "uses-material-design: true".\n'
+      ),
     );
     expect(processManager, hasNoRemainingExpectations);
   });

